File: frontend/src/lib/wifi-security.ts
import type {
  WirelessAuthCipher,
  WirelessAuthType,
  WirelessPmf,
} from "@/lib/api"

// Which Wi-Fi security settings go together (#177). Mirrors
// api/wifi_security.py, which enforces it; here it only narrows the form's
// options so an invalid combination can't be picked. Blank = not documented.

export const SECURITY_MODE_GROUPS: {
  label: string
  options: { value: WirelessAuthType; label: string }[]
}[] = [
  {
    label: "Personal",
    options: [
      { value: "wpa2-personal", label: "WPA2-Personal" },
      { value: "wpa3-personal", label: "WPA3-Personal (SAE)" },
      { value: "wpa2-wpa3-personal", label: "WPA2/WPA3-Personal" },
    ],
  },
  {
    label: "Enterprise",
    options: [
      { value: "wpa2-enterprise", label: "WPA2-Enterprise" },
      { value: "wpa3-enterprise", label: "WPA3-Enterprise" },
      { value: "wpa2-wpa3-enterprise", label: "WPA2/WPA3-Enterprise" },
    ],
  },
  {
    label: "Open",
    options: [
      { value: "owe", label: "Enhanced Open (OWE)" },
      { value: "open", label: "Open" },
    ],
  },
  {
    label: "Legacy",
    options: [
      { value: "wpa-personal", label: "WPA Personal (PSK)" },
      { value: "wpa-enterprise", label: "WPA Enterprise" },
      { value: "wep", label: "WEP (legacy)" },
    ],
  },
]

export const CIPHER_LABEL: Record<Exclude<WirelessAuthCipher, "">, string> = {
  auto: "Auto",
  aes: "AES-CCMP",
  "gcmp-256": "GCMP-256",
  tkip: "TKIP (legacy)",
}

export const PMF_LABEL: Record<Exclude<WirelessPmf, "">, string> = {
  disabled: "Disabled",
  optional: "Optional",
  required: "Required",
}

type Mode = WirelessAuthType

const WPA12: WirelessAuthCipher[] = ["auto", "tkip", "aes"]
const CIPHERS: Record<Mode, WirelessAuthCipher[]> = {
  "": ["auto", "tkip", "aes", "gcmp-256"],
  open: [],
  owe: ["auto", "aes", "gcmp-256"],
  wep: [],
  "wpa-personal": WPA12,
  "wpa-enterprise": WPA12,
  "wpa2-personal": WPA12,
  "wpa2-enterprise": WPA12,
  "wpa3-personal": ["auto", "aes", "gcmp-256"],
  "wpa3-enterprise": ["auto", "aes", "gcmp-256"],
  "wpa2-wpa3-personal": ["auto", "aes"],
  "wpa2-wpa3-enterprise": ["auto", "aes"],
}

const ALL_PMF: WirelessPmf[] = ["disabled", "optional", "required"]
const PMF: Record<Mode, WirelessPmf[]> = {
  "": ALL_PMF,
  open: ["disabled"],
  wep: ["disabled"],
  owe: ["required"],
  "wpa-personal": ALL_PMF,
  "wpa-enterprise": ALL_PMF,
  "wpa2-personal": ALL_PMF,
  "wpa2-enterprise": ALL_PMF,
  "wpa3-personal": ["required"],
  "wpa3-enterprise": ["required"],
  "wpa2-wpa3-personal": ["optional", "required"],
  "wpa2-wpa3-enterprise": ["optional", "required"],
}

/** Modes that authenticate with a stored passphrase. */
const PSK_MODES = new Set<Mode>([
  "wpa-personal",
  "wpa2-personal",
  "wpa3-personal",
  "wpa2-wpa3-personal",
  "wep",
])

export const ciphersFor = (mode: Mode) => CIPHERS[mode]
export const pmfFor = (mode: Mode) => PMF[mode]
/** Blank (not documented) keeps the passphrase field; a mode decides. */
export const takesPassphrase = (mode: Mode) => !mode || PSK_MODES.has(mode)

File: api/wifi_security.py
"""Which Wi-Fi security settings go together (#177).

A security mode says the framework (WPA2/WPA3, Personal/Enterprise, OWE,
open); the cipher and PMF (Protected Management Frames) have to fit it. The
tables here are the one place that says what fits; the serializer refuses
anything else, and the form offers only what fits (frontend
``lib/wifi-security.ts`` mirrors them). Blank always means "not documented".

The older "WPA Personal" / "WPA Enterprise" values stay valid, as they came
in from NetBox and existing records, meaning WPA or WPA2 unspecified.
"""
from __future__ import annotations

PERSONAL = {"wpa-personal", "wpa2-personal", "wpa3-personal", "wpa2-wpa3-personal"}
ENTERPRISE = {"wpa-enterprise", "wpa2-enterprise", "wpa3-enterprise", "wpa2-wpa3-enterprise"}
#: Modes that authenticate with a passphrase stored as the PSK. SAE (WPA3)
#: still derives from one, and WEP has a key of its own.
PSK_MODES = PERSONAL | {"wep"}

#: mode → the ciphers it can run ("" = not documented).
CIPHERS = {
    "": {"", "auto", "tkip", "aes", "gcmp-256"},
    "open": {""},
    "owe": {"", "auto", "aes", "gcmp-256"},
    "wep": {""},
    "wpa-personal": {"", "auto", "tkip", "aes"},
    "wpa-enterprise": {"", "auto", "tkip", "aes"},
    "wpa2-personal": {"", "auto", "tkip", "aes"},
    "wpa2-enterprise": {"", "auto", "tkip", "aes"},
    # WPA3 forbids TKIP; GCMP-256 is WPA3 (Enterprise 192-bit, and SAE-GCMP).
    "wpa3-personal": {"", "auto", "aes", "gcmp-256"},
    "wpa3-enterprise": {"", "auto", "aes", "gcmp-256"},
    # A transition network serves both: only what both sides share.
    "wpa2-wpa3-personal": {"", "auto", "aes"},
    "wpa2-wpa3-enterprise": {"", "auto", "aes"},
}

#: mode → the PMF settings it allows. WPA3 and OWE require PMF; a
#: transition network cannot switch it off; open and WEP have none.
PMF = {
    "": {"", "disabled", "optional", "required"},
    "open": {"", "disabled"},
    "wep": {"", "disabled"},
    "owe": {"", "required"},
    "wpa-personal": {"", "disabled", "optional", "required"},
    "wpa-enterprise": {"", "disabled", "optional", "required"},
    "wpa2-personal": {"", "disabled", "optional", "required"},
    "wpa2-enterprise": {"", "disabled", "optional", "required"},
    "wpa3-personal": {"", "required"},
    "wpa3-enterprise": {"", "required"},
    "wpa2-wpa3-personal": {"", "optional", "required"},
    "wpa2-wpa3-enterprise": {"", "optional", "required"},
}

_CIPHER_NAME = {"auto": "Auto", "tkip": "TKIP", "aes": "AES-CCMP", "gcmp-256": "GCMP-256"}


def problems(mode: str, cipher: str, pmf: str, *, psk: bool) -> dict[str, str]:
    """Field → what is wrong with this combination; empty when it fits.
    ``psk`` is whether the network has (or is being given) a passphrase."""
    mode, cipher, pmf = mode or "", cipher or "", pmf or ""
    out = {}
    if cipher not in CIPHERS.get(mode, CIPHERS[""]):
        out["auth_cipher"] = (
            f"{_CIPHER_NAME.get(cipher, cipher)} doesn't go with this security mode."
            if cipher else "Pick a cipher that goes with this security mode."
        )
    if pmf not in PMF.get(mode, PMF[""]):
        allowed = sorted(p for p in PMF.get(mode, PMF[""]) if p)
        out["pmf"] = f"This security mode allows PMF {' or '.join(allowed)}."
    if psk and mode and mode not in PSK_MODES:
        out["psk"] = ("This security mode takes no passphrase - clear the stored "
                      "one to switch to it.")
    return out
